### DISCUSSION TITLE
Using selected or assigned relationship data in forms as default value

### DISCUSSION THREAD

**Discussion by GlitchWitch** [Budibase/budibase#4807] [Help] [answered]
I am trying to create a default value for a form text field that pulls data from other form fields.
I am using the following handlebars expression:
```
{{ Form.Fields.clients }} -- {{ Form.Fields.Project Type }} - {{ date Form.Fields.Start Date "MM-YYYY"}}
````

The `Project Type` and `Start Date` both appear correctly, however the `clients` field from the relationship picker does not.

I have also tried using `{{ Form.Fields.clients._primaryDisplay }}` but without any luck.

See the below gif for a demonstration.
![](https://user-images.githubusercontent.com/32590246/156971952-a87f34c7-5999-48bd-98ba-c7d85a4ae43f.gif)

_What am I missing to make this work?_

**Comment by gerhardmpl** [comment, marked as answer]
The `ro_ta_55*` is an internal Budibase DB id (`_id`). Can you give `{{ Form.Fields.clients.0.primaryDisplay }}` a try? You can have a look at  #4018 which is not exactly what you are asking for, but related. Make sure that the intended column is set to be the primary display column.

**Reply by GlitchWitch** [reply]
Using `{{ Form.Fields.clients.0.primaryDisplay }}` causes the rendered text within the `Title` default value to be blank in that area. The `Title` column is set as the table display column.

![](https://user-images.githubusercontent.com/32590246/157009710-0d21a0a1-8f5a-46da-a3cb-fc7edca8eadb.gif)

**Reply by mjashanks** [reply]
I've just tried using the format `{{ Form.Fields.clients.0.primaryDisplay }} -- {{ Form.Fields.Project Type }}` and it works of for me.

I noticed from the gif, that your title gets set to `{{ [ .....].[ clients] ....}} ` when we see those square brackets, it usually means that the binding is broken. Sometimes that can happen when you remove a component that you are binding to (even if you re-add it). Its probably worth deleting the binding and building it up again to try.

Happy to have a look if you want to send an export of your app here, or to support@budibase.com

**Reply by GlitchWitch** [reply]
Deleting and re-adding the binding did fix the display a bit, thanks for that tip! 

However `Form.Fields.clients.0.primaryDisplay ` still remains blank. To confirm the Client dropdown shown in the gif is a relationship selector and the field is in fact `clients`. 

I've sent an email with the export. Thanks :smile:

**Reply by mjashanks** [reply]
Got it. I see the issue now. Apparently, the relationship picker works differently on new-forms than it does on existing ones.

On "new" forms, the selected value of the relationship is simply the `_id` - rather than an object like `{ primaryDisplay: "..", _id: "..." }`

I got this to work, with a bit of effort... 

1. Put a `Clients` data provider around the whole form.
2. In the `Default Value` field, I ran the javascript below, which searches the `Clients` data provider for the selected client record:

```javascript
const client = $("Clients.Rows").find(c => c._id == $("Form.Fields.clients"));

let clientName = "";
if(client) {
	clientName = client.Name;
}

const startDate = new Date($("Form.Fields.Start Date"));

return `${clientName} - ${$("Form.Fields.Project Type")} - ${startDate.getMonth() + 1}-${startDate.getFullYear()}`;
```

**Reply by GlitchWitch** [reply]
What do you mean when you talk about "new" forms? As in budibase changed the way that forms are made recently? Is there a way to use the seemingly simpler old form?

I tried your suggestion but I am getting `Error while executing JS` along with `Empty string passed to getElementById(). 16` in console.log.

To confirm I...
 1. Created a new data provider, set the data dropdown to `Clients`
 2. Placed the existing `Form` directly below/within the `New Data Provider`
 3. Set the `Default Value` to the JS provided

**Reply by mjashanks** [reply]
By "New" forms I mean a form used for creating a new record, instead of updating an existing one.

`$("Clients.Rows")` refers to a dataprovider named "Clients" - you might just have to rename your data provider.

**Reply by GlitchWitch** [reply]
You are right, I needed to rename my data provider to match and that fixed it! Thank you, I really appreciate your help once again @mjashanks!

**Comment by GlitchWitch** [comment]
Tangentially related —and do let me know if I should make a separate thread for this— but is it possible to set a relationship dropdown's default value based on the current user's assigned relationship? For example, a client fills out the same form as described earlier, can we hide the `Form.Fields.clients` if it's being filled by the client and set it to something like `{{ Current User.client }}`? I tried this and various permutations such as `{{ Current User.client.0._id }}` and  `{{ Current User.client.0.primaryDisplay }}` as the `default value` but did not have any success.

**Reply by mjashanks** [reply]
I think it should be possible, you might need to use an options picker instead of a relationship picker. I describe how to do this in here: https://github.com/Budibase/budibase/discussions/4579

**Reply by GlitchWitch** [reply]
Thanks @mjashanks 

I tried the following
1. Add new `Data Provider` titled `Clients` with data set to `Clients` table
2. Placed the existing Form directly below/within the new `Clients` Data Provider
3. Delete the exiting relationship picker
4. Created a new `Options Picker` with the field `clients` (same as the previous relationship picker)
5. Set the options source to `Data Provider` and the options provider to `Clients`, Label Column `Name`, Value Column `_id`
.. So far this works...
6. Tried the default values of `{{ Current User.client }}`, `{{ Current User.client.0._id }}`, and `{{ Current User.client.0.primaryDisplay }}` however the default option does not change accordingly. Manually selecting a value however does result in the row saving correctly.

**Reply by mjashanks** [reply]
Copying and editing the JS i wrote above - see below. I've added a few lines at the top that will use the current user's client, if nothing is selected in the picker.

```javascript

let clientId = $("Form.Fields.clients");
if (!clientId) {
  clientId = $("Current User.client.0._id");
}
const client = $("Clients.Rows").find(c => c._id == clientId);

let clientName = "";
if(client) {
   clientName = client.Name;
}

const startDate = new Date($("Form.Fields.Start Date"));

return `${clientName} - ${$("Form.Fields.Project Type")} - ${startDate.getMonth() + 1}-${startDate.getFullYear()}`;
```
